File: src/condash/pty.py
"""PTY session lifecycle for the embedded terminal.

A PTY's lifetime is decoupled from any single WebSocket: a tab refresh
should detach the viewer without killing the shell. Each session lives in
the per-process registry on :attr:`AppState.pty_sessions`, keyed by an
opaque session id; the WebSocket attaches and detaches via
:func:`attach_ws`, while :func:`pump_session` keeps draining the master fd
into a per-session ring buffer (and the live viewer when one is attached).

The shutdown path (:func:`reap_all`) sends SIGTERM to every live shell so
``ui.run`` can exit cleanly without orphaning ports.
"""

from __future__ import annotations

import asyncio
import fcntl
import json
import logging
import os
import secrets
import signal
import struct
import sys
import termios
from dataclasses import dataclass, field
from typing import TYPE_CHECKING

from fastapi import WebSocket, WebSocketDisconnect

if TYPE_CHECKING:
    from .config import CondashConfig
    from .state import AppState

log = logging.getLogger(__name__)

# Ring-buffer cap. 256 KiB fits a few screens of scrollback per tab; beyond
# that we trim from the head. Bound so a detached tab that produces fast
# output can't grow memory unboundedly.
_BUFFER_CAP = 256 * 1024


@dataclass
class PtySession:
    """Server-side pty + its ring buffer. Decoupled from any WebSocket."""

    session_id: str
    pid: int
    fd: int
    shell: str
    cwd: str
    cols: int = 80
    rows: int = 24
    buffer: bytearray = field(default_factory=bytearray)
    out_queue: asyncio.Queue = field(default_factory=asyncio.Queue)
    attached_ws: WebSocket | None = None
    pump_task: asyncio.Task | None = None


def resolve_terminal_shell(cfg: CondashConfig) -> str:
    """Single source of truth for which shell the pty actually launches.

    Priority: explicit ``terminal.shell`` config → ``$SHELL`` env → /bin/bash.
    """
    if cfg.terminal.shell:
        return cfg.terminal.shell
    return os.environ.get("SHELL") or "/bin/bash"


async def spawn_session(
    state: AppState,
    *,
    override_cwd: str | None = None,
    use_launcher: bool = False,
) -> PtySession | None:
    """Fork a new shell in a pty, register it on ``state``, and start its pump.

    Child starts cwd'd at ``override_cwd`` (when supplied and the directory
    exists), else ``conception_path`` (else ``$HOME``) with ``TERM=
    xterm-256color`` and ``-l`` so login rc-files run. ``override_cwd``
    must already be sandbox-validated by the caller — this function trusts
    it and does not re-check.

    When ``use_launcher`` is set, the child execs
    ``terminal.launcher_command`` (shlex-parsed) directly instead of a
    login shell. ``session.shell`` is reported as the launcher's argv[0]
    so the client-side chip label still has something meaningful. When
    that process exits the ws closes and the tab disappears — matching
    what happens when the user types ``exit`` in a regular shell.
    """
    import pty
    import shlex

    cfg = state.cfg
    argv: list[str]
    shell_label: str
    if use_launcher:
        raw_command = cfg.terminal.launcher_command if cfg is not None else ""
        if not raw_command or not raw_command.strip():
            log.warning("term: launcher requested but terminal.launcher_command is empty")
            return None
        try:
            argv = shlex.split(raw_command)
        except ValueError as exc:
            log.warning("term: malformed launcher_command %r: %s", raw_command, exc)
            return None
        if not argv:
            return None
        shell_label = argv[0]
    else:
        shell_label = (
            resolve_terminal_shell(cfg)
            if cfg is not None
            else os.environ.get("SHELL") or "/bin/bash"
        )
        argv = [shell_label, "-l"]

    ctx = state.get_ctx()
    if override_cwd and os.path.isdir(override_cwd):
        cwd = override_cwd
    else:
        cwd = str(ctx.base_dir) if ctx.base_dir.is_dir() else os.path.expanduser("~")

    pid, fd = pty.fork()
    if pid == 0:
        # Child: cwd, env, then exec. os._exit on failure so we don't run
        # parent-only asyncio finally clauses.
        try:
            os.chdir(cwd)
        except OSError:
            pass
        os.environ["TERM"] = "xterm-256color"
        try:
            os.execvp(argv[0], argv)
        except OSError:
            os._exit(127)

    # Parent: wire the pty fd into the asyncio event loop.
    fl = fcntl.fcntl(fd, fcntl.F_GETFL)
    fcntl.fcntl(fd, fcntl.F_SETFL, fl | os.O_NONBLOCK)

    session = PtySession(
        session_id=secrets.token_urlsafe(8),
        pid=pid,
        fd=fd,
        shell=shell_label,
        cwd=cwd,
    )

    loop = asyncio.get_running_loop()

    def _on_readable() -> None:
        try:
            data = os.read(fd, 65536)
        except BlockingIOError:
            return
        except OSError:
            data = b""
        if not data:
            session.out_queue.put_nowait(None)
            try:
                loop.remove_reader(fd)
            except (OSError, ValueError):
                pass
            return
        session.out_queue.put_nowait(data)

    loop.add_reader(fd, _on_readable)
    session.pump_task = asyncio.create_task(pump_session(state, session))
    state.pty_sessions[session.session_id] = session
    return session


async def pump_session(state: AppState, session: PtySession) -> None:
    """Drain the pty's read queue into the ring buffer + any attached ws.

    Runs for the pty's entire lifetime. On EOF (shell exited) unregisters
    the session, sends an ``exit`` frame to whoever was viewing, and
    reaps the child.

    Coalesces any chunks already sitting in the queue into one ws frame
    before awaiting the socket. A large paste floods the pty with echo
    output; ``os.read(fd, 4096)`` reads those in 4 KiB chunks but our
    reader callback loops the queue up fast, so by the time we ``await``
    on ``out_queue.get()`` there are often several pending chunks. Sending
    one frame per chunk forces the client to render (and xterm to parse)
    in 4 KiB increments; coalescing cuts frame count ~10× for the common
    case and makes large paste echo feel instant.
    """
    while True:
        data = await session.out_queue.get()
        if data is None:
            # EOF — shell exited. Tear the session down.
            state.pty_sessions.pop(session.session_id, None)
            ws = session.attached_ws
            session.attached_ws = None
            if ws is not None:
                try:
                    await ws.send_text(json.dumps({"type": "exit"}))
                except (WebSocketDisconnect, RuntimeError, OSError):
                    pass
                try:
                    await ws.close()
                except (WebSocketDisconnect, RuntimeError, OSError):
                    pass
            try:
                os.close(session.fd)
            except OSError:
                pass
            try:
                os.waitpid(session.pid, os.WNOHANG)
            except ChildProcessError:
                pass
            return

        chunks = [data]
        eof_pending = False
        while not session.out_queue.empty():
            try:
                extra = session.out_queue.get_nowait()
            except asyncio.QueueEmpty:
                break
            if extra is None:
                eof_pending = True
                break
            chunks.append(extra)
        data = b"".join(chunks) if len(chunks) > 1 else chunks[0]

        # Append to ring buffer, trimming the head once we overshoot the
        # cap. `del buffer[:n]` on a bytearray is O(n) but cheap at these
        # sizes (256 KiB) and only runs when the buffer is actually full.
        session.buffer.extend(data)
        overflow = len(session.buffer) - _BUFFER_CAP
        if overflow > 0:
            del session.buffer[:overflow]

        ws = session.attached_ws
        if ws is not None:
            try:
                await ws.send_bytes(data)
            except (WebSocketDisconnect, RuntimeError, OSError):
                # Viewer went away (e.g. F5). Detach so the next attach
                # replays from the buffer; keep pty running.
                if session.attached_ws is ws:
                    session.attached_ws = None

        if eof_pending:
            session.out_queue.put_nowait(None)


async def write_all(fd: int, data: bytes) -> bool:
    """Write ``data`` to ``fd`` in full, yielding on EAGAIN.

    The pty master is non-blocking (``O_NONBLOCK``), so a single
    ``os.write`` can return fewer bytes than requested — or raise
    ``BlockingIOError`` once the kernel's tty buffer fills up. That's what
    happens on a large paste: the shell drains the buffer far slower than
    a WebSocket frame can deliver, so the first ``os.write`` ships ~64 KiB
    and the rest would be silently dropped or (worse) would surface as an
    ``OSError`` that tore the ws down. Here we loop, registering a writer
    callback on EAGAIN so the event loop wakes us when the fd is writable
    again. Returns ``True`` on success, ``False`` if the fd went bad.
    """
    loop = asyncio.get_running_loop()
    view = memoryview(data)
    while view:
        try:
            written = os.write(fd, view)
        except BlockingIOError:
            fut: asyncio.Future[None] = loop.create_future()

            def _signal() -> None:
                if not fut.done():
                    fut.set_result(None)

            try:
                loop.add_writer(fd, _signal)
            except (OSError, ValueError):
                return False
            try:
                await fut
            finally:
                try:
                    loop.remove_writer(fd)
                except (OSError, ValueError):
                    pass
            continue
        except OSError:
            return False
        if written <= 0:
            # Shouldn't happen on a healthy fd, but guard against an
            # infinite loop if the kernel ever returns 0.
            return False
        view = view[written:]
    return True


async def attach_ws(session: PtySession, ws: WebSocket) -> None:
    """Run the receive loop for a ws that is viewing ``session``.

    Input bytes are written to the pty; resize frames relay TIOCSWINSZ.
    On disconnect we only clear ``attached_ws`` — the pty keeps running,
    output keeps going into the ring buffer, ready for the next attach.
    """
    try:
        while True:
            try:
                msg = await ws.receive()
            except WebSocketDisconnect:
                break
            mtype = msg.get("type")
            if mtype == "websocket.disconnect":
                break
            if msg.get("bytes"):
                ok = await write_all(session.fd, msg["bytes"])
                if not ok:
                    break
                continue
            text = msg.get("text")
            if not text:
                continue
            try:
                obj = json.loads(text)
            except ValueError:
                continue
            if obj.get("type") == "resize":
                cols = max(2, int(obj.get("cols") or 80))
                rows = max(2, int(obj.get("rows") or 24))
                session.cols, session.rows = cols, rows
                try:
                    fcntl.ioctl(
                        session.fd,
                        termios.TIOCSWINSZ,
                        struct.pack("HHHH", rows, cols, 0, 0),
                    )
                except OSError:
                    pass
    except (WebSocketDisconnect, RuntimeError, OSError) as exc:
        log.debug("attach_ws: receive loop ended: %s", exc)
    finally:
        # Detach only — pty stays alive for the next reconnect.
        if session.attached_ws is ws:
            session.attached_ws = None


def reap_all(state: AppState) -> None:
    """SIGTERM every live pty. Called on server shutdown."""
    for session in list(state.pty_sessions.values()):
        try:
            os.kill(session.pid, signal.SIGTERM)
        except ProcessLookupError:
            pass
        except OSError:
            pass
    state.pty_sessions.clear()


def supports_pty() -> bool:
    """``pty.fork`` works on Linux + macOS only (Windows would need ConPTY)."""
    return sys.platform in ("linux", "darwin")

File: src-tauri/src/pty.rs
//! PTY session lifecycle for the embedded terminal.
//!
//! A PTY's lifetime is decoupled from any single WebSocket: a tab refresh
//! should detach the viewer without killing the shell. Each session lives
//! in the per-process [`PtyRegistry`], keyed by an opaque session id; the
//! WebSocket attaches and detaches via [`PtySession::attach_ws`] /
//! [`PtySession::detach_ws`], while the pump task drains the PTY master
//! into a ring buffer (and forwards to the attached viewer when one is
//! present).
//!
//! Linux + macOS only for now — `portable-pty` would pick ConPTY on
//! Windows but the launcher/exec flow here isn't wired to it.

use std::collections::HashMap;
use std::io::{Read, Write};
use std::path::PathBuf;
use std::sync::{Arc, Mutex};

use portable_pty::{native_pty_system, Child, CommandBuilder, MasterPty, PtySize};
use rand::Rng;
use tokio::sync::mpsc;

/// Cap on the per-session scrollback ring buffer. 256 KiB — enough for a
/// few screens of output. Larger pastes trim from the head. Matches
/// Python's `_BUFFER_CAP`.
pub const BUFFER_CAP: usize = 256 * 1024;

/// Bytes emitted by the PTY pump. `Data(b)` is raw output; `Exit` is
/// "the shell is gone, reap the session".
#[derive(Debug)]
pub enum PumpMessage {
    Data(Vec<u8>),
    Exit,
}

/// Server-side PTY session. Decoupled from any WebSocket viewer.
///
/// All mutable state (buffer, attached_ws marker, cols/rows) lives behind
/// a `Mutex` since the pump task runs on a blocking thread and the
/// WebSocket receive loop runs on tokio.
pub struct PtySession {
    pub session_id: String,
    pub shell: String,
    pub cwd: PathBuf,
    /// Shared ring buffer — scrollback the pump task writes into and the
    /// WebSocket attach path reads on first connect.
    pub buffer: Arc<Mutex<Vec<u8>>>,
    /// Last-observed size. Used when a viewer resizes *after* the
    /// session is created.
    pub size: Arc<Mutex<PtySize>>,
    /// Writer end of the PTY (master side). Used to send user input +
    /// resize ioctls.
    writer: Arc<Mutex<Box<dyn Write + Send>>>,
    /// Master PTY — kept so we can issue resize() on it without dropping
    /// the file descriptor.
    master: Arc<Mutex<Box<dyn MasterPty + Send>>>,
    /// Child-process handle — used by `kill()` on shutdown.
    child: Arc<Mutex<Box<dyn Child + Send + Sync>>>,
    /// `Some(tx)` when a viewer is attached; the pump task forwards
    /// chunks to it until the tx is dropped or `None`.
    viewer: Arc<Mutex<Option<mpsc::UnboundedSender<PumpMessage>>>>,
}

impl PtySession {
    pub fn cols(&self) -> u16 {
        self.size.lock().expect("size mutex").cols
    }
    pub fn rows(&self) -> u16 {
        self.size.lock().expect("size mutex").rows
    }
    pub fn snapshot_buffer(&self) -> Vec<u8> {
        self.buffer.lock().expect("buffer mutex").clone()
    }
    /// Attach a viewer to this session. Replaces any existing viewer's
    /// sender (the old one is dropped so its receiver ends, mirroring
    /// Python's "displace the old viewer" behaviour).
    pub fn attach_viewer(&self) -> mpsc::UnboundedReceiver<PumpMessage> {
        let (tx, rx) = mpsc::unbounded_channel();
        // Push the buffer first so the new viewer sees scrollback.
        let replay = self.snapshot_buffer();
        if !replay.is_empty() {
            let _ = tx.send(PumpMessage::Data(replay));
        }
        let mut guard = self.viewer.lock().expect("viewer mutex");
        *guard = Some(tx);
        rx
    }

    /// Detach the viewer — the pump keeps running and buffering output.
    pub fn detach_viewer(&self) {
        let mut guard = self.viewer.lock().expect("viewer mutex");
        *guard = None;
    }

    /// Write user input to the PTY master.
    pub fn write_input(&self, bytes: &[u8]) -> std::io::Result<()> {
        let mut w = self.writer.lock().expect("writer mutex");
        w.write_all(bytes)?;
        w.flush()
    }

    /// Resize the PTY. Clamps to `>=2` rows/cols: many TUIs crash on 0/1.
    pub fn resize(&self, cols: u16, rows: u16) {
        let cols = cols.max(2);
        let rows = rows.max(2);
        let new_size = PtySize {
            rows,
            cols,
            pixel_width: 0,
            pixel_height: 0,
        };
        *self.size.lock().expect("size mutex") = new_size;
        if let Ok(master) = self.master.lock() {
            let _ = master.resize(new_size);
        }
    }

    /// SIGTERM the child — used by [`PtyRegistry::reap_all`].
    pub fn kill(&self) {
        if let Ok(mut child) = self.child.lock() {
            let _ = child.kill();
        }
    }
}

/// Per-process registry of live PTY sessions — cloneable handle shared
/// by the HTTP routes and the shutdown reaper.
#[derive(Clone, Default)]
pub struct PtyRegistry {
    inner: Arc<Mutex<HashMap<String, Arc<PtySession>>>>,
}

impl PtyRegistry {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn get(&self, session_id: &str) -> Option<Arc<PtySession>> {
        self.inner
            .lock()
            .expect("registry mutex")
            .get(session_id)
            .cloned()
    }

    pub fn insert(&self, session: Arc<PtySession>) {
        self.inner
            .lock()
            .expect("registry mutex")
            .insert(session.session_id.clone(), session);
    }

    pub fn remove(&self, session_id: &str) -> Option<Arc<PtySession>> {
        self.inner
            .lock()
            .expect("registry mutex")
            .remove(session_id)
    }

    pub fn len(&self) -> usize {
        self.inner.lock().expect("registry mutex").len()
    }

    pub fn is_empty(&self) -> bool {
        self.len() == 0
    }

    /// SIGTERM every live PTY and drop them from the registry. Called
    /// on server shutdown.
    pub fn reap_all(&self) {
        let mut guard = self.inner.lock().expect("registry mutex");
        for (_, session) in guard.iter() {
            session.kill();
        }
        guard.clear();
    }
}

/// What to execute inside the PTY.
pub enum SpawnMode {
    /// Login shell (`<shell> -l`) — the normal `/ws/term` flow.
    LoginShell { shell: String },
    /// Arbitrary argv parsed from `terminal.launcher_command` — mirrors
    /// Python's `use_launcher=True` path.
    Launcher { argv: Vec<String> },
    /// `<shell> -lc <template-with-path-replaced>` — the runner flow
    /// (Phase 4 slice 2). Kept on this enum so a single spawn path
    /// covers every caller.
    RunnerCommand {
        shell: String,
        template: String,
        path: String,
    },
}

/// Spawn a PTY running the requested command, register it, and start
/// the pump task on `spawn_blocking`. The pump drains PTY master output
/// into the ring buffer (and any attached viewer) until the child exits.
///
/// `cwd` is the directory to start the child in — must already be
/// sandbox-validated by the caller.
pub fn spawn_session(
    registry: &PtyRegistry,
    mode: SpawnMode,
    cwd: PathBuf,
    cols: u16,
    rows: u16,
) -> std::io::Result<Arc<PtySession>> {
    let pty_system = native_pty_system();
    let size = PtySize {
        rows,
        cols,
        pixel_width: 0,
        pixel_height: 0,
    };
    let pair = pty_system
        .openpty(size)
        .map_err(|e| std::io::Error::new(std::io::ErrorKind::Other, format!("openpty: {e}")))?;

    let (argv, shell_label): (Vec<String>, String) = match mode {
        SpawnMode::LoginShell { shell } => (vec![shell.clone(), "-l".into()], shell),
        SpawnMode::Launcher { argv } => {
            let label = argv.first().cloned().unwrap_or_default();
            (argv, label)
        }
        SpawnMode::RunnerCommand {
            shell,
            template,
            path,
        } => {
            let resolved = template.replace("{path}", &path);
            (vec![shell.clone(), "-lc".into(), resolved], shell)
        }
    };

    if argv.is_empty() {
        return Err(std::io::Error::new(
            std::io::ErrorKind::InvalidInput,
            "empty argv",
        ));
    }

    let mut cmd = CommandBuilder::new(&argv[0]);
    for arg in &argv[1..] {
        cmd.arg(arg);
    }
    cmd.cwd(&cwd);
    cmd.env("TERM", "xterm-256color");

    let child = pair
        .slave
        .spawn_command(cmd)
        .map_err(|e| std::io::Error::new(std::io::ErrorKind::Other, format!("spawn: {e}")))?;
    // Close the slave on the parent side — the child retains its copy.
    drop(pair.slave);

    let reader = pair.master.try_clone_reader().map_err(|e| {
        std::io::Error::new(std::io::ErrorKind::Other, format!("clone reader: {e}"))
    })?;
    let writer = pair
        .master
        .take_writer()
        .map_err(|e| std::io::Error::new(std::io::ErrorKind::Other, format!("take writer: {e}")))?;

    let session_id: String = {
        let mut rng = rand::thread_rng();
        (0..16)
            .map(|_| {
                let n: u8 = rng.gen_range(0..36);
                if n < 10 {
                    (b'0' + n) as char
                } else {
                    (b'a' + (n - 10)) as char
                }
            })
            .collect()
    };

    let session = Arc::new(PtySession {
        session_id: session_id.clone(),
        shell: shell_label,
        cwd,
        buffer: Arc::new(Mutex::new(Vec::with_capacity(4096))),
        size: Arc::new(Mutex::new(size)),
        writer: Arc::new(Mutex::new(writer)),
        master: Arc::new(Mutex::new(pair.master)),
        child: Arc::new(Mutex::new(child)),
        viewer: Arc::new(Mutex::new(None)),
    });

    registry.insert(session.clone());

    // Spawn the pump on a blocking thread — portable-pty's reader is
    // synchronous. The pump drives the ring buffer + forwards to any
    // attached viewer, then emits `Exit` on EOF.
    let pump_session = session.clone();
    let pump_registry = registry.clone();
    std::thread::Builder::new()
        .name(format!("condash-pty-pump-{session_id}"))
        .spawn(move || {
            pump_loop(pump_session, pump_registry, reader);
        })
        .map_err(|e| std::io::Error::new(std::io::ErrorKind::Other, format!("pump thread: {e}")))?;

    Ok(session)
}

/// Synchronous pump — runs on a dedicated thread for the PTY's entire
/// lifetime. Reads up to 64 KiB at a time, appends to the ring buffer
/// (trimming the head when past the cap), and forwards to the attached
/// viewer when one is present. On EOF (the shell exited), fires an
/// `Exit` to the viewer and drops the session from the registry.
fn pump_loop(session: Arc<PtySession>, registry: PtyRegistry, mut reader: Box<dyn Read + Send>) {
    let mut buf = [0u8; 64 * 1024];
    loop {
        match reader.read(&mut buf) {
            Ok(0) => break,
            Ok(n) => {
                let chunk = &buf[..n];
                {
                    let mut ring = session.buffer.lock().expect("buffer mutex");
                    ring.extend_from_slice(chunk);
                    let overflow = ring.len().saturating_sub(BUFFER_CAP);
                    if overflow > 0 {
                        ring.drain(..overflow);
                    }
                }
                let viewer = { session.viewer.lock().expect("viewer mutex").clone() };
                if let Some(tx) = viewer {
                    let _ = tx.send(PumpMessage::Data(chunk.to_vec()));
                }
            }
            Err(e) if e.kind() == std::io::ErrorKind::Interrupted => continue,
            Err(_) => break,
        }
    }
    // EOF or read error — tell the viewer and unregister.
    let viewer = { session.viewer.lock().expect("viewer mutex").take() };
    if let Some(tx) = viewer {
        let _ = tx.send(PumpMessage::Exit);
    }
    registry.remove(&session.session_id);
}

/// `pty.supports_pty()` equivalent — we support Linux + macOS. Windows
/// would need a separate spawn path (ConPTY), so the /ws/term handler
/// returns an early error there.
pub fn supports_pty() -> bool {
    cfg!(any(target_os = "linux", target_os = "macos"))
}

/// Resolve the shell to launch for `/ws/term`. Priority:
/// 1. Explicit `override_shell` from the config layer (Phase 4 slice 2).
/// 2. `$SHELL` environment variable.
/// 3. `/bin/bash`.
pub fn resolve_terminal_shell(override_shell: Option<&str>) -> String {
    if let Some(s) = override_shell {
        let trimmed = s.trim();
        if !trimmed.is_empty() {
            return trimmed.into();
        }
    }
    std::env::var("SHELL").unwrap_or_else(|_| "/bin/bash".into())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn resolve_terminal_shell_priority() {
        assert_eq!(resolve_terminal_shell(Some("/bin/zsh")), "/bin/zsh");
        // Empty / whitespace override falls through.
        std::env::set_var("SHELL", "/bin/override-sh");
        assert_eq!(resolve_terminal_shell(Some("   ")), "/bin/override-sh");
        std::env::remove_var("SHELL");
        assert_eq!(resolve_terminal_shell(None), "/bin/bash");
    }

    #[test]
    fn supports_pty_matches_target() {
        let got = supports_pty();
        if cfg!(target_os = "linux") || cfg!(target_os = "macos") {
            assert!(got);
        } else {
            assert!(!got);
        }
    }

    #[cfg(target_os = "linux")]
    #[test]
    fn spawn_pty_echoes_and_exits() {
        // Drive a tiny shell command end-to-end: spawn, collect output,
        // observe exit. Uses sh -c rather than bash for robustness.
        let reg = PtyRegistry::new();
        let mode = SpawnMode::Launcher {
            argv: vec!["/bin/sh".into(), "-c".into(), "printf hello; exit 0".into()],
        };
        let session =
            spawn_session(&reg, mode, std::env::temp_dir(), 80, 24).expect("spawn session");

        // Attach a viewer so the pump forwards chunks in real time (and
        // the initial buffer replay is empty).
        let mut rx = session.attach_viewer();

        // Drain until Exit or timeout.
        let timeout_at = std::time::Instant::now() + std::time::Duration::from_secs(5);
        let mut output = Vec::new();
        let mut exited = false;
        while std::time::Instant::now() < timeout_at {
            match rx.try_recv() {
                Ok(PumpMessage::Data(bytes)) => output.extend_from_slice(&bytes),
                Ok(PumpMessage::Exit) => {
                    exited = true;
                    break;
                }
                Err(tokio::sync::mpsc::error::TryRecvError::Empty) => {
                    std::thread::sleep(std::time::Duration::from_millis(20));
                }
                Err(_) => break,
            }
        }
        assert!(exited, "pump never emitted Exit; output={output:?}");
        let text = String::from_utf8_lossy(&output);
        assert!(
            text.contains("hello"),
            "expected 'hello' in PTY output, got: {text:?}"
        );
        // Session should have been auto-removed by the pump on EOF.
        assert!(
            reg.get(&session.session_id).is_none(),
            "session not removed after exit"
        );
    }
}
